guidedmodules/app_source_connections.py

```diff
@@ -0,0 +1,611 @@
+    def __init__(self, repo, path, user, pw):
+        # Create client.
+        from github import Github
+        g = Github(user, pw)
+        self.repo = g.get_repo(repo)
+        self.path = (path or "") + "/"
+
+        # Run a quick call to check access.
+        from github.GithubException import GithubException
+        try:
+            self.repo.get_dir_contents(self.path)
+        except GithubException as e:
+            raise fs.errors.CreateFailed(e.data.get("message"))
+
+    def scandir(self, path, namespaces=None, page=None):
+        from fs.info import Info
+        for cf in self.repo.get_dir_contents(self.path + path):
+            yield Info({
+                "basic": {
+                    "name": cf.name,
+                    "is_dir": cf.type == "dir",
+                },
+                "hash": {
+                    "sha1": cf.sha,
+                }
+            })
+
+    def openbin(self, path, mode="r", **options):
+        if mode not in ("r", "rb"): raise ValueError("Invalid open mode. Must be 'r' or 'rb'.")
+        import base64, io
+        cf = self.repo.get_contents(self.path + path)
+        if cf.type != "file": raise ValueError("path is a directory")
+        if cf.encoding != "base64": raise ValueError("content encoding is unrecognized")
+        content = base64.b64decode(cf.content)
+        return io.BytesIO(content)
+
+
+# This class implements AppSourceConnection using the GitHub API,
+# using the GithubApiFilesystem class above to wrap the GitHub API
+# in a PyFilesystem2 filesystem. options should look like:
+# {
+#	"type": "github",
+#   "repo": "organization/repository",
+#   "path": "apps", # optional
+#   "auth": { "user": "username", "pw": "password" } # optional
+# }
+class GithubApiAppSourceConnection(PyFsAppSourceConnection):
+    def __init__(self, source, options):
+        if not isinstance(options.get("repo"), str): raise ValueError("The AppSource is misconfigured: missing or invalid 'repo'.")
+        if not isinstance(options.get("path"), (str, type(None))): raise ValueError("The AppSource is misconfigured: missing or invalid 'path'.")
+        if not (isinstance(options.get("auth"), dict) and isinstance(options["auth"].get("user"), str)): raise ValueError("The AppSource is misconfigured: missing or invalid 'auth.user'.")
+        if not (isinstance(options.get("auth"), dict) and isinstance(options["auth"].get("pw"), str)): raise ValueError("The AppSource is misconfigured: missing or invalid 'auth.pw'.")
+        super().__init__(source, lambda : GithubApiFilesystem(
+            options["repo"], options.get("path"),
+            options["auth"]["user"], options["auth"]["pw"]))
+
+
+class GitRepositoryFilesystem(SimplifiedReadonlyFilesystem):
+    def __init__(self, url, branch, path, ssh_key=None):
+        self.url = url
+        self.branch = branch or None
+        self.path = (path or "") + "/"
+        self.ssh_key = ssh_key
+
+        # Create a local git working directory.
```

Comment:
Maybe if we added here " # Create a local git working directory that will be automatically destroyed after use."